### DISCUSSION TITLE
[Amplitued 연동 길드] amplitude util 함수 interface를 GA와 비슷하게 만들어야 하는가?

### DISCUSSION THREAD

**Discussion by p-iknow** [EveryAnalytics/react-analytics-provider#127] [General]
## 관련 PR 
https://github.com/EveryAnalytics/react-analytics-provider/issues/105

## Discussion 
 amplitude util 함수 interface를 GA와 비슷하게 만들어야 하는가?  더 구체적으로 하자면 `googleanalytics.pagview()` 가 존재한다면, `amplitude.pageview()` 도 있어야 하는가? 
 
 ## 논의의 배경 
 - @milooy: amplitude util 도 ga와 동일하게 만들어야 하지 않을까? 
 - @p-iknow: ga에는 공식적으로 pageview 라는 event가 존재하지만, amplitude에서는 그런 구분이 없어요. 그럼에도 불구하고 동일하게 만들어야 할까?) 
 또한 [gtag](https://developers.google.com/gtagjs/reference/api) 와 [amplitude](https://amplitude.github.io/Amplitude-JavaScript/) 각가 api의 규격이 달라서 공통적으로 빼기는 힘들듯
 
## **제안하는 방향성**

- GA, Amplitude 각 라이브러리 기본 API에 대한 헬퍼만 제공하자.
- e.g.
    - 기본
        - googleAnalytics.event(...) :  ga의 initialize스크립트를 쉽게 쓸 수 있는 헬퍼
        - googleAnalytics.event(...) // :man-gesturing-ok: ga의 event란 api를 쉽게 쓸 수 있는 헬퍼
    - 기본이 아닌 헬퍼
        - googleAnalytics.click(...): ga는 click이란 api가 없다.
        - googleAnalytics.modal(...): ga는 modal이란 api가 없다.

## 위 제안의 이유

- 지금 만들고 있는 util 함수들을 너무 공급자 위주로 만들고 있다. util 함수들은 사용자의 필요에 의해서 만들어져야 한다. 우리가 만들고 있는 util 함수들이 얼마나 쓰일지, 얼마나 필요한지 우리는 잘 모른채 만든다.
- 우리가 만든 ga 나 amplitude util 함수의 기본적인 기능만 있어도 react-analytics-provider를 쓰는데 큰 지장이 없다.
- 아주 기본적인 util를 제공한 뒤 beta open을 하고, 사용자로 부터 피드백을 얻어 개발 방향성을 얻어 가는 방식이 더 에자일 하다.
- util 함수를 많이 만들면 그 만큼 api 에 대한 문서화 등의 유지보수가 많이 들어간다. 초기에는 기본적인 util만 만들고, 관련 api 문서를 ga, amplitude 원래 문서에 위임하면, 유지보수 비용이 덜 들어간다.

## **방향성에 공감한 뒤 정해야할 것**

- 무엇이 기본적인 util인가, GA, Amplitude 각자가 다른 api를 제공하므로 기본적인 util은 GA, Amplitude에서 각자 정한다.
- 기존에 만들어둔 함수들이 위 기준(기본적인)에 해당하지 않으면 잠정적 삭제

**Comment by MiryangJung** [comment]
궁금한게 있습니다! amplitude에는 pageview라는 구분이 없다면, `amplitude.pageview()` 는 어떻게 만드는 건가요...?

**Reply by p-iknow** [reply]
^ interface를 맞추기 위해 아래와 같은 식으로 만들어야 하지 않을까? 라고 생각했습니다. 
`amplitude.pageview = (customPageviewEventName = 'pageView') => amplitude.logEvent(customPageviewEventName);`

@MiryangJung

**Comment by hiphapis** [comment]
제 의견은, 추상화된 interface는 동일하게 제공을 하는게 맞다고 생각합니다.

예를들어 pageview('/url/path') 라고 하면
GA 설정이 되어 있으면 ga.pageview를 실행시키고
Ampt 설정이 되어 있으면 ampt.pageview도 실행이 되는거죠.

자 여기서 ampt 에서 pageview라는 액션이 없다면, 우리가 만들면 되는거죠!

----

> 더 구체적으로 하자면 googleanalytics.pagview() 가 존재한다면, amplitude.pageview() 도 있어야 하는가?

- 추상화된 interface라면 동일하게 가는것이고,
- ow level에서 이루어지는거라면 달라질 수 도 있겠지만,  가급적이면  같은 네이밍이 좋을 듯 합니다. 그게 컨벤션이 되는거니깐요.

**Comment by milooy** [comment]
AnalyticsProvider에서는 외부 라이브러리 디펜던시 없이
우리의 의지만큼 추상화 할 수 있지만
(e.g. onPageView, onClick, onModal, onImpression 등)

실제 외부 라이브러리의 헬퍼인 utility함수에선
우리가 어느 선까지 개념을 '창조'(ㅎㅎ) 해도 될지 불분명해서 이뤄지는 문제로 보이네요.

실제로 우리가 만든 googleAnalytics.click 도 GA에는 없는 개념이죠 🤔 
(googleAnalytics의 이벤트에 파라미터를 추가해서 래핑했을뿐)

만약 amplitude.click 개념을 만들거라면 amplitude.pageView도 만들 수 있지 않을까요?

**Comment by milooy** [comment]
구두 논의:
@p-iknow @MiryangJung 님의 고견을 듣다가... 없는 개념 창조에 컨선이 들기 시작했고
amplitude.click 를 만들지 않을거면 ga.click도 없애야하지 않을까? 란 생각에 가까워졌습니다 ㅎㅎ
영창님 논의정리 해주시면 보고 더 이야기 나눠요~

영창님께서 정리해서 올려주실거예요 기대기대
